core/lib/src/fairing/ad_hoc.rs

```diff
@@ -153,17 +153,20 @@ impl Fairing for AdHoc {
         Info { name: self.name, kind }
     }
 
-    fn on_attach(&self, rocket: Rocket) -> Result<Rocket, Rocket> {
+    async fn on_attach(&self, rocket: Rocket) -> Result<Rocket, Rocket> {
         if let AdHocKind::Attach(ref mutex) = self.kind {
-            let mut opt = mutex.lock().expect("AdHoc::Attach lock");
-            let f = opt.take().expect("internal error: `on_attach` one-call invariant broken");
-            f(rocket)
+            let f = mutex
```

Comment:
Would you prefer it to look more like the previous version, or to go further and remove this `let` binding?

Reply:
So, this is actually necessary so that `opt` (`MutexGuard`) is not in scope across an `await` point. The cleaner-looking `drop(opt)` before the `await` doesn't work due to rust-lang/rust#57478.

This is the other way that works:

```rust
            let f = {
                let mut opt = mutex.lock().expect("AdHoc::Attach lock");
                opt.take().expect("internal error: `on_attach` one-call invariant broken")
            };
            f(rocket).await
```